pkg/scheduler/framework/plugins/volumezone/volume_zone.go

```diff
@@ -111,63 +112,132 @@ func (pl *VolumeZone) Filter(ctx context.Context, _ *framework.CycleState, pod *
-			if err != nil {
-				klog.InfoS("Failed to parse label, ignoring the label", "label", fmt.Sprintf("%s:%s", k, v), "err", err)
-				continue
+		for _, key := range topologyLabels {
+			if value, ok := pv.ObjectMeta.Labels[key]; ok {
+				volumeVSet, err := volumehelpers.LabelZonesToSet(value)
+				if err != nil {
+					klog.InfoS("Failed to parse label, ignoring the label", "label", fmt.Sprintf("%s:%s", key, value), "err", err)
+					continue
+				}
+				podPVTopologies = append(podPVTopologies, pvTopology{
+					pvName: pv.Name,
+					key:    key,
+					values: volumeVSet,
+				})
 			}
+		}
+	}
+	return podPVTopologies, nil
+}
 
-			if !volumeVSet.Has(nodeV) {
-				klog.V(10).InfoS("Won't schedule pod onto node due to volume (mismatch on label key)", "pod", klog.KObj(pod), "node", klog.KObj(node), "PV", klog.KRef("", pvName), "PVLabelKey", k)
-				return framework.NewStatus(framework.UnschedulableAndUnresolvable, ErrReasonConflict)
-			}
+// PreFilterExtensions returns prefilter extensions, pod add and remove.
+func (pl *VolumeZone) PreFilterExtensions() framework.PreFilterExtensions {
+	return nil
+}
+
+// Filter invoked at the filter extension point.
+//
+// It evaluates if a pod can fit due to the volumes it requests, given
+// that some volumes may have zone scheduling constraints.  The requirement is that any
+// volume zone-labels must match the equivalent zone-labels on the node.  It is OK for
+// the node to have more zone-label constraints (for example, a hypothetical replicated
+// volume might allow region-wide access)
+//
+// Currently this is only supported with PersistentVolumeClaims, and looks to the labels
+// only on the bound PersistentVolume.
+//
+// Working with volumes declared inline in the pod specification (i.e. not
+// using a PersistentVolume) is likely to be harder, as it would require
+// determining the zone of a volume during scheduling, and that is likely to
+// require calling out to the cloud provider.  It seems that we are moving away
+// from inline volume declarations anyway.
+func (pl *VolumeZone) Filter(ctx context.Context, cs *framework.CycleState, pod *v1.Pod, nodeInfo *framework.NodeInfo) *framework.Status {
+	// If a pod doesn't have any volume attached to it, the predicate will always be true.
+	// Thus we make a fast path for it, to avoid unnecessary computations in this case.
+	if len(pod.Spec.Volumes) == 0 {
+		return nil
+	}
+	node := nodeInfo.Node()
+	if node == nil {
+		return framework.NewStatus(framework.Error, "node not found")
+	}
+
+	var podPVTopologies []pvTopology
+
+	state, err := getStateData(cs)
+	if err != nil {
+		//Fallback to calculate pv list here
+		var status *framework.Status
+		podPVTopologies, status = pl.getPVbyPod(ctx, pod)
+		if !status.IsSuccess() {
+			return status
+		}
+	} else {
+		podPVTopologies = state.podPVTopologies
+	}
+
+	for _, pvTopology := range podPVTopologies {
+		v, ok := node.Labels[pvTopology.key]
+		if !ok {
+			continue
+		}
```

Comment:
This behavior is not exactly the same as before.

Before this PR, a node would pass the Filter if it didn't have any of the topology labels. But if it had, let's say, zone but not region, and the PV had zone and region, the node wouldn't pass the filter.

Do you agree?

@msau42 can you think why it was implemented like that? Was it a bug?

Reply:
> But if it had, let's say, zone but not region, and the PV had zone and region, the node wouldn't pass the filter.

That's true, I think this is a little confusing, as if a node has a zone label matched but doesn't have a region label, it can't pass the filter. However, if a node dones't have either a zone label or a region label, it can pass the filter.

IMO, the former node is more qualified, since it has a label that meets the requirements and the latter has none😂

Reply:
I think the correct behavior should be that it doesn't pass the Filter. But I'll let @msau42 confirm.

We might have to preserve the behavior if there is a possibility that the pods would run. If they would fail at startup, I think we are safe to change the behavior at scheduling.

Reply:
I believe this is to handle a single-zone cluster scenario where the node may not have any topology labels. I'm not sure if it's possible to encounter this with any of our built-in cloud providers, but if someone has their own node creation process it may be possible to encounter this scenario. 

It is possible for this to still succeed (a volume plugin could have some fallback behavior to handle single zone), and especially now that we have PV.NodeAffinity, the zone/region labels have more of an informational purpose (although nothing is stopping anyone from relying on this old behavior).

Reply:
If the cluster has a single zone, shouldn't the PVs also have no zone label?

But it looks like we could be breaking someone. @kidddddddddddddddddddddd, please make sure to treat nodes without topology labels specially. And add a unit test for it, if there's none already.

Reply:
> But if it had, let's say, zone but not region, and the PV had zone and region, the node wouldn't pass the filter.

@msau42 so in this scenario, we should fail the filter, and node with no topology labels is a special scenario where we let it pass the filter?

Reply:
> please make sure to treat nodes without topology labels specially. And add a unit test for it, if there's none already.

I think there is a test case for nodes without topology labels. I'm going to create a case for the scenario we talked before(a node has a zone label matched but doesn't have a region label can't pass the filter.)